src/topology/config/component.rs

```diff
@@ -0,0 +1,75 @@
+use inventory;
+use serde::{Deserialize, Serialize};
+use std::marker::PhantomData;
+use toml::Value;
+
+/// Describes a component plugin storing its type name, an example config, and
+/// other useful information about the plugin.
+pub struct ComponentDescription<T: Sized> {
+    pub type_str: &'static str,
+    example_value: fn() -> Result<Value, String>,
```

Comment:
Why a `fn` here instead of just a `toml::Value`? I'm also curious why it would be fallible.

Reply:
I figured it would be preferred over storing a serialized example for all components in memory, but I'm not attached so I'm happy to pre-bake it if you prefer.

Since the fallback way of generating an example is to parse an empty string into the type there's a potential for erroring out due to missing fields. I'm pretty sure serde doesn't have a way of enforcing the presence of a field during deserialize so I don't think it's currently possible to fail at that point, but I figured I'd propagate it just in case so we can expose it.

Reply:
Do you think that in the long term these will alway be autogenerated? Or would we eventually provide examples when adding new components as a part of documenting them?

If fallible autogeneration is the intent, I can see the argument for keeping a `Result` to show any problems. Otherwise, I think it'd make sense to reduce this to just an `Option<toml::Value>`. That way it would be present for components that have a good default or provided their own example, and simply not present for those that don't.

Reply:
My intention is for all configs to at least implement `Default`, and in some cases where the config structure is more complex provide a curated example. I doubt I'll manage to get all configs on `Default` right now so we need an interim but I can live with `Option`. I'm not opposed to simply returning an empty `Value::table` either.

I might also rename `new` and `new_with_default` to `new_without_default` and `new` respectively. Seems more progressive to shame the non-defaulters. We can then add `new_with_example` for configs with a curated example.

Reply:
I've updated to return an `Option<Value>`, there's not really much point even attempting the parse of empty strings since I don't think there are any configs that would allow it but I've left it in for now.

Eventually I'll start adding curated config examples, we can support that by adding builder pattern methods to `ComponentDescription<T>` with things like `.with_example` and maybe `.with_summary`.

Reply:
Sounds great!